Add time-rs as optional time backend

tyhi opened the pull request:
This is a preference change (and a breaking one at that) and if you would like to stay on chrono this pr can be closed. 

Performance impact on running cargo bench seems to be about 13%(+-3%) slower on generating and about -10% faster on parsing. While reducing the number of dependencies pulled in (up to 7 less on linux). Benchmarks were ran on Azure using a  Standard D8as_v4 (8 vcpus, 32 GiB memory) vm on Centos 8.4,

The datetime function uses an expect, however chrono uses an expect internally so behavior is exactly the same.

Commit:
Swap chrono for time

huxi commented:
Just a quick feedback since I don't have the time right now: I'm not ignoring you!

I wouldn't merge the PR as it is but we could try to do this in a way that works with either `time` or `chrono`. That way it wouldn't necessarily have to be a breaking change.

tyhi committed:
Change to enable to add time-rs as optional time backend

This will allow a user to use time-rs instead of chrono as the time backend, while keeping chrono as the default to prevent any breaking changes.

tyhi commented:
I agree, and should've done that from the start.

I've renamed the PR and change to use features to keep chrono as the default and keeping all function signatures the same. I've added a `compile_error!` to error on enabling both features though not sure if there's a better way to handle this. 

Let me know if you find anything that needs to be changed, thanks!

huxi commented:
Sorry for ghosting you again. Life is kind of hard right now...

I'd prefer if you could change the code so it compiles even if both `chrono` and `time` are present at the same time.

There's a [justfile](https://github.com/huxi/rusty_ulid/blob/master/justfile) that builds the different combinations and I think it should be possible to enable all optional dependencies. It would be nice if you could extend it and make sure that `just all` still works.

huxi commented on src/lib.rs:
```diff
@@ -283,7 +299,7 @@ impl Ulid {
     ///
     /// Panics if called after `+10889-08-02T05:31:50.655Z`.
-    #[cfg(all(feature = "rand", feature = "chrono"))]
+    #[cfg(all(feature = "rand", feature = "time"))]
```
Still missing the `any`

tyhi replied:
Fixed

huxi commented on src/lib.rs:
```diff
@@ -313,7 +329,7 @@ impl Ulid {
     ///
     /// Panics if called after `+10889-08-02T05:31:50.655Z`.
-    #[cfg(all(feature = "rand", feature = "chrono"))]
+    #[cfg(all(feature = "rand", feature = "time"))]
```
Still missing the `any`

tyhi replied:
Fixed

huxi commented on src/lib.rs:
````diff
@@ -707,7 +723,6 @@ impl Ulid {
     /// # Ok::<(), rusty_ulid::DecodingError>(())
     /// ```
     #[cfg(feature = "chrono")]
-    #[must_use]
````
Why was this removed?

tyhi replied:
Not super sure how that got removed, re-added.

huxi commented on src/lib.rs:
````diff
@@ -716,6 +731,26 @@ impl Ulid {
+    /// # Ok::<(), rusty_ulid::DecodingError>(())
+    /// ```
+    #[cfg(feature = "time")]
+    pub fn datetime(&self) -> OffsetDateTime {
````
Similarly, I'd suggest to have two different methods here. One that returns the chrono type (as before) and one that returns the time type. I'm just struggling how to name it... offsetdatetime?

tyhi replied:
offsetdatetime makes sense and keeps it simple, works for me.

tyhi commented:
No worries on the delay, definitely understandable.

So with latest commit, I performed all the changes, along with changing readme to ignore the code block (for cargo test) and saying that if both features are enabled chrono will be used over time.

I edited the justfile however I'm not super familiar with it so if it needs to be changed let me know. 

Thanks for reviewing the PR so far though!

tyhi committed:
Allow both time backend features to be enabled.

* Re-add removed [must_use]
* Renamed time function to offsetdatetime
* Reflected readme to explain when both backends are enabled
* Added ignore to readme codeblock to stop test from trying to run.

huxi commented:
I just published 1.0.0 including your changes. Had to fix a minor issue in case of both chrono and time but thanks a lot for your contribution!

tyhi commented:
> I just published 1.0.0 including your changes. Had to fix a minor issue in case of both chrono and time but thanks a lot for your contribution!

Thank you for getting this merged and the assistance you provided me!
